Split out finding layerID to be able to add to TextIndicator.

megangardner opened the pull request:
#### 4a8f78b66188e9b86a6599308fb7070b733d06e9
<pre>
Split out finding layerID to be able to add to TextIndicator.
<a href="https://bugs.webkit.org/show_bug.cgi?id=293723">https://bugs.webkit.org/show_bug.cgi?id=293723</a>
<a href="https://rdar.apple.com/152207015">rdar://152207015</a>

Reviewed by Wenson Hsieh.

We would like to use this same logic to place the
textIndicator in the layer tree, so separate it
out so that we can work towards that goal.

* Source/WebCore/editing/Editing.cpp:
(WebCore::computeEnclosingLayer):
* Source/WebCore/editing/Editing.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::computeEnclosingLayerID const):

Canonical link: <a href="https://commits.webkit.org/295586@main">https://commits.webkit.org/295586@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c4beeb317a092310bedc557c7bc6c086af65bc94

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105535 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25280 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15674 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110732 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/56176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107576 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25716 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33785 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80146 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/forms/constraints/infinite_backtracking.html (failure)") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/56176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108541 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20072 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95236 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60455 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/19820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13327 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55569 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/89527 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13368 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113529 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32674 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/24095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89223 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33037 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91466 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88884 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22663 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33758 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/11561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28123 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32600 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/38004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32353 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35702 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33944 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -6115,79 +6115,13 @@ static VisiblePositionRange constrainRangeToSelection(const VisiblePositionRange
-        return;
-
-    if (!endLayer)
+    if (!selection.range())
```
Nit - you could store `selection.range()` in a local variable here to avoid calling it twice.
```suggestion
    auto selectionRange = selection.range();
    if (!selectionRange)
        return;
```

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -6115,79 +6115,13 @@ static VisiblePositionRange constrainRangeToSelection(const VisiblePositionRange
-        state.visualData->enclosingLayerID = WTFMove(identifier);
-        break;
-    }
+    CheckedPtr<RenderLayer> enclosingLayer = layerInformation.enclosingLayer;
```
```suggestion
    CheckedPtr enclosingLayer = layerInformation.enclosingLayer;
```

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -6115,79 +6115,13 @@ static VisiblePositionRange constrainRangeToSelection(const VisiblePositionRange
-
-            if (backing->isFrameLayerWithTiledBacking())
-                return backing->parentForSublayers();
+    EnclosingLayerInfomation layerInformation = computeEnclosingLayer(state.isContentEditable, selection.range().value());
```
Nit - `auto`:

```suggestion
    auto layerInformation = computeEnclosingLayer(state.isContentEditable, selection.range().value());
```

or, you could consider destructuring assignment:

```suggestion
    auto [startLayer, endLayer, enclosingLayer, enclosingGraphicsLayerID] = computeEnclosingLayer(state.isContentEditable, *selectionRange);
```

whsieh commented on Source/WebCore/editing/Editing.cpp:
```diff
@@ -1499,6 +1502,83 @@ SimpleRange adjustToVisuallyContiguousRange(const SimpleRange& range)
+        return { nullptr, nullptr, nullptr, { } };
+
+    if (!isContentEditable && selection.isCaret())
+        return { nullptr, nullptr, nullptr, { } };
```
It's a bit weird to pass in a `range` here that's been converted from a `VisibleSelection`, right before converting the range _back_ into a `VisibleSelection`.

You could consider either changing this to take a `VisibleSelection`, or keep the range and use `positionsForRange` to get the `start` and `end`, and then `isEditablePosition` on either the `start` or `end` to check if it's editable. You'd then change the `selection.isCaret()` check to use `range.collapsed()` instead.

whsieh reviewed (commented):
Looks good — just a few small comments.

whsieh commented on Source/WebCore/editing/Editing.cpp:
```diff
@@ -1499,6 +1502,83 @@ SimpleRange adjustToVisuallyContiguousRange(const SimpleRange& range)
+        return { nullptr, nullptr, nullptr, { } };
+
+    if (!endLayer)
+        return { nullptr, nullptr, nullptr, { } };
```
This might also look a bit cleaner:

```suggestion
        return { };

    if (!endLayer)
        return { };
```

whsieh commented on Source/WebCore/editing/Editing.cpp:
```diff
@@ -1499,6 +1502,83 @@ SimpleRange adjustToVisuallyContiguousRange(const SimpleRange& range)
+
+        return { startLayer, endLayer, layer, identifier };
+    }
+    return { nullptr, nullptr, nullptr, { } };
```
Ditto —

```suggestion
    return { };
```

whsieh commented on Source/WebCore/editing/Editing.h:
```diff
@@ -34,11 +35,13 @@
 namespace WebCore {
 
 class Document;
+class EditorState;
```
This shouldn't be here.

```suggestion
```

whsieh reviewed (approved):
r=me with comments addressed

megangardner committed:
Split out finding layerID to be able to add to TextIndicator.
https://bugs.webkit.org/show_bug.cgi?id=293723
rdar://152207015

Reviewed by Wenson Hsieh.

We would like to use this same logic to place the
textIndicator in the layer tree, so separate it
out so that we can work towards that goal.

* Source/WebCore/editing/Editing.cpp:
(WebCore::computeEnclosingLayer):
* Source/WebCore/editing/Editing.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::computeEnclosingLayerID const):

Canonical link: https://commits.webkit.org/295586@main
